utils/kubernetes/expose/expose.go

```diff
@@ -0,0 +1,521 @@
+	labelsMap    map[string]string
+	protocolsMap map[string]string
+	portsSlice   []string
+	Config
+}
+
+// Expose exposes the given kubernetes resource
+func Expose(
+	clientSet *kubernetes.Clientset,
+	restConfig rest.Config,
+	config Config,
+	resources []Resource,
+) ([]*v1.Service, error) {
+	// continueOnError not only controls if the traversal should continue even after errors
+	continueOnError := true
+
+	tr := Traverser{
+		Client:    clientSet,
+		Resources: resources,
+		Logger:    config.Logger,
+	}
+	createdSvc, err := tr.Visit(func(info Object, err error) (*v1.Service, error) {
+		if config.Namespace == "" {
+			config.Namespace = info.GetNamespace()
+		}
+		if err != nil {
+			return nil, nil
+		}
+
+		// Check if the resource can be exposed or not
+		gk := info.GetObjectKind().GroupVersionKind().GroupKind()
+		if err := canBeExposed(gk); err != nil {
+			return nil, ErrResourceCannotBeExposed(err, gk.Kind)
+		}
+
+		if len(config.Name) > validation.DNS1035LabelMaxLength {
+			config.Name = config.Name[:validation.DNS1035LabelMaxLength]
+		}
+
+		// Map for selectors of the current object
+		selectorsMap, err := mapBasedSelectorForObject(info)
+		if err != nil {
+			return nil, ErrSelectorBasedMap(err)
+		}
+
+		isHeadlessService := config.ClusterIP == "None"
+
+		// protocolsMap stores the protocols for the current object
+		protocolsMap, err := protocolsForObject(info)
+		if err != nil {
+			return nil, ErrProtocolBasedMap(err)
+		}
+
+		// protocolsMap stores the protocols for the current object
+		labelsMap, err := meta.NewAccessor().Labels(info)
+		if err != nil {
+			return nil, ErrLableBasedMap(err)
+		}
+
+		ports, err := portsForObject(info)
+		if err != nil {
+			return nil, ErrPortParsing(err)
+		}
+		if len(ports) == 0 && !isHeadlessService {
+			return nil, ErrPortParsing(fmt.Errorf("no ports found for the non headless resource"))
+		}
+
+		service, err := generateService(serviceConfig{
+			selectorsMap: selectorsMap,
+			protocolsMap: protocolsMap,
+			labelsMap:    labelsMap,
+			portsSlice:   ports,
+			Config:       config,
+		})
+		if err != nil {
+			return nil, ErrGenerateService(err)
+		}
+		config.Logger.Info(fmt.Sprintf("Generated service object %s in namespace %s", service.Name, service.Namespace))
+		helper, err := constructObject(clientSet, restConfig, service)
+		if err != nil {
+			return nil, ErrConstructingRestHelper(err)
+		}
+
+		_, err = helper.Create(config.Namespace, false, service)
+		config.Logger.Info("Service deployed")
+		if err != nil {
+			return nil, ErrCreatingService(err)
+		}
+
+		return service, nil
+	}, continueOnError)
+
+	// This err is already generated
+	// from ErrTraverser
+	return createdSvc, err
+}
+
+func generateService(serviceConfig serviceConfig) (*corev1.Service, error) {
+	ports := []corev1.ServicePort{}
+	for i, port := range serviceConfig.portsSlice {
+		// We can expect the port to be a valid UNIX port and hence
+		// should not cause integer overflow. Hence,
```

Comment:
`Hence` ?

Reply:
@mgfeller actually I was trying to justify the use of `#nosec` here which is in the next line.